rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+
+   ```
+   version \d*;
+   ```
+
+   This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).
+
+   The syntax is open for bikeshedding.
+   Alternatives should be very short and self-describing.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Short and fairly self-descriptive
+   - (+) Not an invasive change
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - `use v\d*;`
+     - (+) Shorter
+     - (-) Does not explain much
+   - `Nix language version \d*`
+     - (+) Says it all
+     - (-) Very long
+   - `with import <language> \d*;`
+     - (+) Allows for forward compatibilty hacks such as better error messages
+     - (-) Will likely mislead beginners to think this is has the same semantics as the original `with import ...;`
+
+   </details>
+
+1. As an exception, the language version declaration is optional for Nix expressions passed directly as an argument to the evaluator and on the REPL, and can be specified with a separate parameter to the evaluator.
+   If no version is specified in a bare Nix expression, assume the most recent language version supported by the evaluator.
+
+   This is to ease use of the REPL and evaluating ad hoc expressions.
+
+   <details><summary>Alternatives</summary>
+
+   - Make it mandatory
+       - (-) This will be very inconvenient to use in the REPL
+
+   - Assume the evaluator's current version
+       - (-) When the evaluator advances in language version, evaluation may fail on existing code
+       - (-) Defeats the purpose of explicit versioning: Which evaluator to use for a given file is left unspecified
+           - (-) Following the latest evaluator version may inadvertently break the code for older evaluators
+       - (+) Don't have to look up the latest version of the Nix language when writing code
+       - (+) Does not clutter the file names for what is supposed to be the latest version of the code
+
+   </details>
+
+1. Language versions prior to 6 are not supported.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Does not require additional development effort
+   - (+) Prior langauge versions are not fully supported by current code already, and the rest of this proposal argues to deprecate old versions in the future in order to keep the implementation manageable
+   - (-) Legacy code will not get support for managing compatibility
+
+   </details>
+
+1. The Nix language evaluator provides a command to output the most recent Nix language version.
+   This command provides options to list all Nix language versions supported by the evaluator.
+
+   <details><summary>Arguments</summary>
+
+   - (+) This is for convenience to determine which features are available
+
+   </details>
+
+1. `builtins.langVersion` returns the language version used for evaluating the given expression.
```

Comment:
This needs to be made more precise w.r.t. its interaction with cross-version interoperability.

Reply:
Good point. What should the following evaluate to?

```
# b.nix
builtins.langVersion
```

```nix
# b.nix
version 7;
import a.nix == builtins.langVersion
```

This whole part is actually underspecified. What would be the purpose of even having `langVersion` if the version is determined for the entire expression already?

Reply:
I have the same question. I can't think of a case where an expression would want to know what language **syntax** version it is being evaluated as. The version will be static based on the version declared in the file.

I think that `builtins.langVersion` is more `builtins.builtinLibVersion` and should probably preserve its current behaviour of being global. Leave the version declaration defined here to be purely for syntax changes.

This is probably also a good simplification as `builtins` can be the same object globally. The downside would be that we can never rename anything in `builtins` or change the behaviour of a function. But that seems like less useful anyways as we can always add new functions.

Reply:
> This is probably also a good simplification as `builtins` can be the same object globally.

By globally you mean across versions? Possibly. In any case, the idea of versioning `builtins` separately is worth considering.

Reply:
If you want cross-version interoperability the way you describe below, then `builtins` indeed needs to be versioned and your example above would evaluate to `false`. And I think this is correct, because the point of doing this versioning in the first place is to make sure code of an older version evaluates exactly the same way as it always did, even in the presence of breaking language changes.

Reply:
> By globally you mean across versions?

Yes, I mean across all files in a single evaluation even if they use different syntax versions.

I think it is definitely worth considering as I'm not sure it makes sense to make breaking changes in `builtins` anyways. 

1. `builtins` can be handled by runtime logic such as an assert or a fallback. It doesn't have the same strictness as syntax changes do where the code won't run at all if the parse fails.
2. We only really need to do this if we want to make breaking changes to `builtins` such as changing the behaviour of an existing function. I'm not sure we want to do that anyways as it is likely to cause bugs for people who didn't notice the change, whereas breaking syntax changes will cause errors at parse time even if the code isn't run. (Unless we make changes that affect the behaviour of old code without causing it to fail parsing which is also something that I think we should try to avoid for this reason).

So that means that the only real thing that matching `builtins` to the syntax version would allow is removal of deprecated functions. But we would still need to keep them in Nix for older versions, so the only real benefit is that it can help people avoid bugs by pushing them off of "bad" functions. But the downside is that this push is only caught at runtime, not compile time.

The downsides would be that we aren't exposing new functions to older syntax versions, forcing an artificial update to these users (or do we add new functions all the way back but only remove/change based on the syntax flag?)

TL;DR since changes to `builtins` can be handled at runtime and can only be detected at runtime I don't think we necessarily want to tie them to the same version flag.

Reply:
> I'm not sure it makes sense to make breaking changes in builtins anyways.

Counter point: I'd love to see `toString` fixed one day.

Reply:
I would too, but I'm not sure that I would like to see it changed throughout my project without me knowing. I'd probably prefer something like `toString2`. Then in a decade we can remove `toString` and in another decade alias `toString2` as `toString`.

Maybe this could be handled better if we treat `builtins` as magic. So you can't do `x = builtins` and accesses to `builtins.toString` are checked as a syntax error. (We can also allow things like `builtins.newFunc ? fallback`). That way we could remove the old `toString` at the same time we add a new `toString2` or `toStr` and know that users need to make the explicit update. That way as soon as you say `version 8;` you will get a syntax error that you are using `builtins.toString` (with a helpful error message about how it is removed and you should consider `toStr` instead and here is a document describing the differences).

I guess we could publish a migration guide and expect every user to read it carefully before updating the version statement in their code, but I don't have much faith that will happen. I think we would need reliable auto-migration or errors on the new version for changes to make the process work well. (Except for possibly exception circumstances where it is worth the risk of breakage).

Reply:
I doubt that any amount of documentation will avoid people introducing bugs during a version upgrade. If a compatibility change doesn't raise an error reliably people will miss it and it will bite them in the future.

For example in Rust editions you can see that all off the changes cause compile errors in previous versions. You can update the edition you are using, then make changes until your code compiles, and your code will work afterwards. I think this is a very good workflow and we should aim towards it. Expecting people to carefully read and understand the changelog and then audit their code is a recipe for bugs. I think that is best avoided.

I think we can do things like error in one version, then accept in the next, and tell people that they shouldn't skip versions when upgrading. But I think that anything that goes from working -> working with changed behaviour is very risky.

Reply:
Resolved, removed `builtins.langVersion` from the proposal as it's not related to the goals of this RFC.